### PR TITLE
Init estimation for regression.

### PR DESCRIPTION
**Pull request by trivialfis**

### CODE REVIEW

**Comment by trivialfis** [comment]
Some small refactors are coupled in this PR to cleanly split up CPU impl and CUDA impl for utilities.

File: src/objective/regression_loss.h [hunk 1]
```diff
@@ -39,6 +52,15 @@ struct LinearSquareLoss {
 
   static const char* Name() { return "reg:squarederror"; }
   static ObjInfo Info() { return {ObjInfo::kRegression, true, false}; }
+
+  static bool InitEstimation(Context const* ctx, MetaInfo const& info,
+                             linalg::Tensor<float, 1>* base_margin) {
+    CheckInitInputs(info);
+    base_margin->Reshape(1);
+    auto out = base_margin->HostView();
+    out(0) = WeightedMean(ctx, info);
```
Comment (RAMitchell) [inline src/objective/regression_loss.h:61]: If you wanted to make a more general implementation that would work for the other loss functions, maybe you could just calculate the gradient/hessian, sum them up and calculate base margin as w=G/H.

Comment (trivialfis) [reply src/objective/regression_loss.h:61]: We don't have prediction at this stage, as a result there's no gradient. This initialization is to establish the first prediction.

Comment (trivialfis) [reply src/objective/regression_loss.h:61]: @RAMitchell I believe the scale pos(itive) weight is exclusively for logistic, and I agree that we can remove it.

Comment (RAMitchell) [reply src/objective/regression_loss.h:61]: You could just predict 0 and get the gradients.

**Comment by trivialfis** [comment]
I will change the init code to all objectives in followups.

**Comment by trivialfis** [comment]
Will extract a small portion of this PR if I can't get everything working in time.

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**